SCons/Job.py

```diff
@@ -396,19 +406,47 @@ def start(self):
                     else:
                         if task.needs_execute():
                             # dispatch task
-                            self.tp.put(task)
+                            self.thread_pool.put(task)
                             jobs += 1
+                            retired -= 1
+                            if retired == 0:
+                                break
                         else:
                             task.executed()
                             task.postprocess()
 
-                if not task and not jobs: break
+                # If 'task' is true-ish, then we enqueued a new task,
+                # otherwise we didn't. If we didn't enqueue a task and
+                # we don't have any jobs pending, then there is
+                # nothing left to do.
+                if not (task or jobs):
+                    break
+
+                # If we couldn't find any new tasks requiring
+                # execution, we need to wait for a completion.
+                block = (task is None)
 
-                # Let any/all completed tasks finish up before we go
-                # back and put the next batch of tasks on the queue.
-                while True:
-                    task, ok = self.tp.get()
-                    jobs -= 1
+                # If we have reached our quota for outstanding jobs,
+                # we must wait for at least one job to complete before
+                # re-entering the loop above.
+                block |= (jobs == self.maxjobs)
+
+                retired = 0
+                while jobs:
+                    try:
+                        task, ok = self.thread_pool.get(block=block)
+                        jobs -= 1
+                        retired += 1
+                    except queue.Empty:
+                        # We did a non-blocking wait, but there was
+                        # nothing ready yet. See if we can find another
+                        # task to enqueue. If we can't we will end up
+                        # back in this loop but we will do a blocking
+                        # wait.
+                        break
```

Comment:
So this is the only exit from this loop now?
(Vs before it would also check if results queue was empty).

Reply:
Which will only happen when block = False, which is only set if no new tasks were added and we're not already at max jobs.

(not counting users hit's control-c or similar)

Reply:
Right, only way out is to hit `break` here, which only happens on the non-blocking case. But, if you did a blocking wait, you will come back around and do a non-blocking one, since we unconditionally set `block` to `false` if we don't `break`.
- If you are non-blocking, an empty queue takes you directly to the `break`.
- If you are blocking, your next attempt after unblocking will be non blocking, and an empty queue takes you to `break`.

Reply:
Also, the prior check of whether the results queue was empty is basically equivalent to doing the non-blocking `get`.